### PR TITLE
Internal macro `construct_arguments` now generates separate Args sets

### PR DESCRIPTION
**Pull request by UebelAndre**
`construct_arguments` now generates separate sets of arguments, one for the process wrapper, another to launch `rustc` from the process wrapper, and a final set for Rust flags. This allows rules like clippy to pass flags to the process wrapper and not have this function need to be aware of clippy (making the rules more contained).

### CODE REVIEW

**Review by illicitonion — APPROVED** [review]
I am a little surprised this works, but it seems pretty reasonable considering it does! :)

**Review by jfgoog — COMMENTED** [review]
This seems like a reasonable abstraction, and makes the handling of --touch-file cleaner.

But I don't see how this helps with _error_format, which I think is the bigger problem for me in https://github.com/bazelbuild/rules_rust/pull/848

You can probably also do this much more simply, and without having to modify all the callers of construct_arguments, if you just prepend --touch-file in clippy.bzl:

```
args = ["--touch-file", clippy_marker.path, args]
```

I appreciate your help on this, but I think a complicating factor is that aspect-on-aspect handling in Bazel just doesn't work right at the moment, so it may be best for me to hold off on this until that's been fixed, because I am basically unable to test my changes right now.

**Comment by UebelAndre** [comment]
> You can probably also do this much more simply, and without having to modify all the callers of construct_arguments, if you just prepend --touch-file in clippy.bzl:
> 
> ```
> args = ["--touch-file", clippy_marker.path, args]
> ```

True, but there's only two callers so far and this is a method in a private package so I feel like the impact is minimal.

> But I don't see how this helps with _error_format, which I think is the bigger problem for me in #848

> I appreciate your help on this, but I think a complicating factor is that aspect-on-aspect handling in Bazel just doesn't work right at the moment, so it may be best for me to hold off on this until that's been fixed, because I am basically unable to test my changes right now.

Is there an issue tracking what's broken? Also, don't mean to ping pong you but gonna continue this thread in https://github.com/bazelbuild/rules_rust/pull/848#discussion_r674059185 😅

**Comment by UebelAndre** [comment]
Testing for this would be covered by https://github.com/bazelbuild/rules_rust/pull/851

File: rust/private/rustc.bzl [hunk 1]
```diff
@@ -341,12 +340,16 @@ def construct_arguments(
...
 
     Returns:
         tuple: A tuple of the following items
-            - (Args): An Args object with common Rust flags
+            - (struct): A struct of arguments used to run the `Rustc` action
+                - process_wrapper (Args): Arguments for the process wrapper
+                - launcher (Args): Arguments for invoking rustc via the process wrapper
+                - rustc (Args): Rust flags for the Rust compiler
+                - all (list): A list of all `Args` objects in the order listed above.
+                    This is to be passed to the `arguments` parameter of actions
```
Comment (hlopko) [inline rust/private/rustc.bzl:352]: I wonder if there is naming that will show that the first and the third args contain command line flags, and launcher contains pretty much a path to the rustc binary?

What do you think about:

* process_wrapper_flags (or _opts, probably not _args as it is loaded term)
* rustc_path (launcher is also a bit loaded term, in java/python rules the launcher is a binary that starts the runtime and bundles all the code in a single file)
* rustc_flags (or _opts)

**Comment by hlopko** [comment]
Overall I like this change a lot. Thank you Andre!

File: rust/private/rustc.bzl (hunk 1 above)
Comment (UebelAndre) [reply rust/private/rustc.bzl:352]: Updated 😄 

I, kinda arbitrarily, went with `_flags` but I wonder if there could there be a vocabulary/language guide made for Bazel so decisions like this are clearer across various rules?

**Pull request merged**